projects/dataset_recorder/robot_recorder.h

```diff
@@ -0,0 +1,225 @@
+            LOGD << std::setprecision(10)
+                << "GPS quality: " << gpsQual
+                << " latitude: " << coord.lat.value()
+                << " longitude: " << coord.lon.value()
+                << " num sats: " << gpsData.numberOfSatellites
+                << " time: " << time.value() << std::endl;
+
+            // converting to UTM
+            const auto utm = MapCoordinate::latLonToUTM(coord);
+            m_Recorder.record(utm.toVector(), degree_t{ yaw }, m_Frame,
+                              pitch.value(), roll.value(), botSpeed,
+                              turnAngle.value(), utm.zone,
+                              (int) gpsData.gpsQuality,
+                              millimeter_t{ gpsData.horizontalDilution }.value(),
+                              time.value());
+        } else {
+            /*
+             * No new data was available. Indicate this by writing NaNs to the
+             * CSV file. We can always estimate these missing values post hoc
+             * with interpolation.
+             */
+            m_Recorder.record(Vector3<millimeter_t>::nan(), degree_t{ yaw },
+                              m_Frame, pitch.value(), roll.value(), botSpeed,
+                              turnAngle.value(), "", -1, NAN, time.value());
+        }
+
+        return time;
+    }
+
+    const auto &getGPSData() const
+    {
+        return m_GPSData;
+    }
+
+private:
+    GPS::GPSReader &m_GPS;
+    GPS::GPSData m_GPSData;
+    std::unique_ptr<Video::Input> m_Camera;
+    cv::Mat m_Frame;
+    Navigation::ImageDatabase m_Database;
+    Navigation::ImageDatabase::RouteRecorder<Navigation::ImageDatabase::VideoFileWriter> m_Recorder;
+    Stopwatch m_Stopwatch;
+
+    static std::unique_ptr<Video::Input> getCamera()
+    {
+#ifdef DUMMY_CAMERA
+        LOGW << "USING DUMMY CAMERA!!!!!!!!";
+        return std::make_unique<Video::RandomInput<>>(cv::Size{ 360, 100 });
+#else
+        return Video::getPanoramicCamera();
+#endif
+    }
+
+    units::frequency::hertz_t getFrameRate() const
+    {
+#ifdef DUMMY_CAMERA
+        return units::frequency::hertz_t{ 30 };
+#else
+        return m_Camera->getFrameRate();
+#endif
+    }
+
+    /*
+     * Jetsons sadly don't record the time and date while powered off. We can
+     * sometimes get the date from the GPS though (also sadly) we don't always
+     * get given this information. So let's have a command-line option to force
+     * using the system clock. The onus is then on the user to check that the
+     * system clock's date is correct.
+     *
+     * We do however always get *time* information with GPS coordinates, so we
+     * can at least rely on that being correct.
+     */
+    const std::tm &getCurrentDateTime(bool useSystemClock) const
+    {
+        using namespace std::literals;
+
+        auto currentTime = m_GPS.getCurrentDateTime();
+
+        // We already have current date from GPS
+        if (currentTime.tm_year) {
+            goto out;
```

Comment:
Why not just add:

```c++
std::tuple<int, int, int> getCurrentDate(bool useSystemClock)
{
     if (useSystemClock) {
        LOGW << "Using system clock to get current date. This may not be correct!";
        const auto rawTime = std::time(nullptr);
        const auto &systemTime = *gmtime(&rawTime);
        return std::make_tuple(systemTime.tm_mday, 
                               systemTime.tm_mon,
                               systemTime.tm_year);
    }
    else {
        Stopwatch sw;
        sw.start();
        while (sw.elapsed() < 5s) {
            currentTime = m_GPS.getCurrentDateTime();
            if (currentTime.tm_year) {
                return std::make_tuple(currentTime.tm_mday, 
                                       currentTime.tm_mon,
                                       currentTime.tm_year);
            }
        }

        // ...and if we didn't, give up
        throw std::runtime_error{ "Timed out waiting for date from GPS" };
    }
}
```
and then, in ``getCurrentDateTime``:
```c++
if(!currentTime.tm_year) {
    std::tie(currentTime.tm_mday, 
             currentTime.tm_mon,
             currentTime.tm_year) = getCurrentDate(useSystemClock);

}
```
You could also go a bit less modern C++ and pass ``currentTime`` by reference and set the day, month and year fields directly 🤷

Reply:
The only problem with this is that it doesn't get the timezone info. The timezone will almost certainly be UTC anyway, but we do at least parse that info...

Reply:
Actually I could just ensure that it is UTC with a ``BOB_ASSERT()`` and then do it your cleaner way....

Reply:
It occurs to me that another way around the ``goto`` ugliness is with the immediately invoked lambda idiom again :wink:

Reply:
your choice! I feel especially with the option of the immediately invoked lambda idiom and this being an exception-allowing codebase, there's not many reasons for a goto